Source/messaging/MessageUnit.h

```diff
@@ -786,6 +805,38 @@ namespace WPEFramework {
                     }
                 }
 
+                void Modules(std::vector<string>& modules) const
+                {
+                    if (_channel.IsOpen() == true) {
+                        Core::ProxyType<MetadataFrame> metaDataFrame(Core::ProxyType<MetadataFrame>::Create());
+
+                        uint8_t buffer[1];
+                        Core::FrameType<0> frame(buffer, 1, 1);
+                        Core::FrameType<0>::Writer writer(frame, 0);
+
+                        writer.Number<metadataFrameProtocol>(metadataFrameProtocol::MODULES);
+
+                        metaDataFrame->Parameters().Set(writer.Offset(), buffer);
+
+                        uint32_t result = _channel.Invoke(metaDataFrame, Core::infinite);
+
+                        if (result == Core::ERROR_NONE) {
+                            uint16_t bufferSize = metaDataFrame->Response().Length();
+                            const uint8_t* buffer = metaDataFrame->Response().Value();
+
+                            Core::FrameType<0> frame(const_cast<uint8_t*>(buffer), bufferSize, bufferSize);
+                            Core::FrameType<0>::Reader reader(frame, 0);
+
+                            size_t iterator = reader.Number<size_t>();
+
+                            while(iterator > 0) {
+                                modules.push_back(reader.NullTerminatedText());
```

Comment:
Suggest to build some logic here. If the module is already in the vector<string> modules, do not add it, and sort it (put it in the right location. If you sort it you also now quicker if it is already in there etc.. (see also remark above).
This will avoid a large std::vector<> with a lot of duplicate modules that need to be sorted and removed later on..

Reply:
Like I replied above - I do agree and I will add a logic here, since this way it is much more memory efficient. And I also think we can get rid of sorting altogether - it was only needed to do the erase to remove duplicates. If we want the modules to be sorted in the UI, than we can do that in JavaScript